### PR TITLE
Move and rename volume-based magnetic field map classes 

### PR DESCRIPTION
**Pull request by whokion**
This MR is a ground work to integrate a volume-based magnetic field map for CMSSW and includes 
- move files from test/celeritas/field to src/celeritas/field
- drop CMS from class names as they are generally applicable to any volume based magnetic map field

### CODE REVIEW

**Comment by whokion** [comment]
@sethrj These moved and renamed files are need to be linked from CMSSW and related interfaces (which will be added after this MR). I am not sure whether src/celeritas/field is the right place for these files, so please advise if there is the better way to wrap these into the celeritas library.  Another option may be to put them under src/accel, which does not look like a good option either.

**Review by sethrj — CHANGES_REQUESTED** [review]
One big thing that's missing is the `test` namespace needs to be removed from all the files that are no longer in `test/`.

There's a different level of rigor required between tests and main code, so these classes need to have a few improvements before moving to the main codebase. The following stand out to me from a quick code review:
- FieldMapReader::CMSFieldMapInput should be moved to the .cc file, and probably renamed to `FieldMapRecord`... and is this format just for the unit test or is it used elsewhere?
- Reinterpret-cast for reading raw bytes is not a very safe or portable entry method.
- I think `MagFieldMap` needs to be renamed to `MapFieldParams`, and the `FieldMapInput` doesn't seem like it's in the right place, since it's using host-only std::vector.
- If the inputs are hardcoded to be floats,  maybe we should do the same for the underlying data? Or vice versa? There is an inconsistency.
- `MapField` should use `UniformGridData` rather than a bunch of `static_cast`
- The conversion to tesla should be done during construction of `FieldMapData` rather than at runtime in `MapField`

**Comment by sethrj** [comment]
(P.S. I do think that `celeritas/field` is the right place to move them — but if the reader is for testing only and won't be used by CMS, it should stay in the test directory.)

**Comment by whokion** [comment]
Thanks for all suggestions - will work on them.  Concerned to `FieldMapReader`, it will be used in `AlongStepFieldMapAction` (which will be added in the next MR), so it may be better to stay in celeritas/field - I originally tried to put this in `**app**` as its function is just  loading a user field map (i.e, constructing a field map and  passing it through `AlongStepFactory`), but had a problem passing the device_ref() of the field map to make_along_step_launcher through multiple layers.  So, an idea is to pass `FieldMapParameter` (which contains the name of map file and grip information that are set in `GlobalSetup`) to `AlongStepAction` to build the final map accordingly.  If `MapField` is intended to be used as a general volume-based 2D field map, it may be still better to stay celeritas/field as a helper class so that each end user does not need to re-implement how to read the field map as there are no general user action interface classes yet.  Of course, I am open to other alternatives.

**Comment by whokion** [comment]
@sethrj I need more input for following items which have not addressed yet
- Reinterpret-cast for reading raw bytes is not a very safe or portable entry method.
  As I recall, this was originally `(char*)` which was suggested to change to `reinterpret_cast<char*>`. So, what is your new recommendation here?
- the FieldMapInput doesn't seem like it's in the right place, since it's using host-only std::vector.
  So, are you suggesting to add a separate `FieldMapInput.hh` as it is used in both `FieldMapReader` and `MapFieldParams`?
- MapField should use UniformGridData rather than a bunch of static_cast
  This requires a re-design of the map data structure (it is not a simple array, but has two components (i.e, r and z) at each r-z grid point) - eventually we may try the 2D texture for the map. I would like to keep it as it is for this MR and revisit the better treatment as a separate MR (you or @amandalund may help on this).
- If the inputs are hardcoded to be floats, maybe we should do the same for the underlying data? Or vice versa?
  Yes, there is an (intended) in-consistency to trade off the size of memory and the intrinsic conversion from float to real_type as the map itself may practically be an approximation.  No problem to change value to real_type, which may be a potential memory issue when the size of map is big in real use cases (i.e, doubling the size) - again it is arguable between the number of grids to approximate the field and the size of the total data - i.e, using float to doubling the number of grids to store keeping the same data size.

**Comment by sethrj** [comment]
>   As I recall, this was originally `(char*)` which was suggested to change to `reinterpret_cast<char*>`. So, what is your new recommendation here?

>   So, are you suggesting to add a separate `FieldMapInput.hh` as it is used in both `FieldMapReader` and `MapFieldParams`?

As I mentioned at the meeting, this isn't a good file format since the metadata and data are separate, the raw data casting is unsafe and possibly undefined behavior, and it's a nonstandard data format. Instead, since I assume this is meant to be a dense R*Z array of data, how about you just define an `MapFieldInput` struct and put it in the `MapFieldParams` header:

```c++
struct FieldMapInput
{
    size_type num_z{};
    size_type num_r{};
    real_type delta{};
    std::vector<double> z_values; // [Z][R], units of tesla
    std::vector< double > r_values; // [Z][R], units of tesla
};
```
Then add JSON I/O operations for it, and the "reader" can be replaced in the unit test/demo app with:
```c++
auto input = nlohmann::json::parse(std::ifstream("field.json")).get<FieldMapInput>();
```
This is a good balance between reusability and not forcing a single arbitrary file format.

>   This requires a re-design of the map data structure (it is not a simple array, but has two components (i.e, r and z) at each r-z grid point) - eventually we may try the 2D texture for the map. I would like to keep it as it is for this MR and revisit the better treatment as a separate MR (you or @amandalund may help on this).

The r and z grid points are still uniform, so you should at least be able to use UniformGrid to e.g. map (r, z) -> (ir, iz). I'd just prefer non-test code to have more robust checking, helper functions, and comments than this file has.

>   Yes, there is an (intended) in-consistency to trade off the size of memory and the intrinsic conversion from float to real_type as the map itself may practically be an approximation.  No problem to change value to real_type, which may be a potential memory issue when the size of map is big in real use cases (i.e, doubling the size) - again it is arguable between the number of grids to approximate the field and the size of the total data - i.e, using float to doubling the number of grids to store keeping the same data size.

I think we should do simplicity first, optimization second. It may turn out that having a variable grid size is a more important optimization than using single-precision. Or if it becomes a big issue we could use a floating point compression library such as ZFP.

**Comment by whokion** [comment]
Following the suggestion,
`auto input = nlohmann::json::parse(std::ifstream("field.json")).get<FieldMapInput>();`
I have a compilation error, [json.txt](https://github.com/celeritas-project/celeritas/files/11298380/json.txt).  Any idea?

**Comment by sethrj** [comment]
Goodness, that's an ugly error message. I don't think I can tell what the problem is without seeing the source code. Could you push it and I'll have a go at fixing it?

**Comment by whokion** [comment]
@sethrj I added the follow code snippet (just for testing jason I/O in `Fields.test.cc)` and had the compilation error:
```
#include <nlohmann/json.hpp>
#include <fstream>
...
    struct FieldTestInput
    {
       size_type num_z{};
       size_type num_r{};
       real_type delta{};
       real_type offset{};
       std::vector<double> z_values; // [Z][R], units of tesla                                      
       std::vector<double> r_values; // [Z][R], units of tesla                                      
    };
    auto input = nlohmann::json::parse(std::ifstream("field.json")).get<FieldTestInput>();
```
I see an exactly same error report [here](https://stackoverflow.com/questions/70228105/when-deserializing-a-struct-with-from-json-error-no-matching-function-for-call).

**Comment by sethrj** [comment]
It looks like maybe you're not defining the JSON accessors? With JSON you have to define the serialization yourself. If you just push the latest branch I can save you the trouble and implement it.

**Comment by whokion** [comment]
@sethrj I do not have new commits other than a quick scratch test inside Fields.test.cc described above.  So, can you help to implement what you suggested and improve the branch?  I also notice that the decoupled linear interpolation inside `MapField::operator()(Real3 const& pos) ` has a potential culprit in calculated values of field components as (z, r) goes to the next point at (iz+1, ir+1) of which field values were not used in the algorithm.  So, we should come up with a correct interpolation using all values at 4-points  without a discontinuity at the diagonal corner- I will take care of this if you do not have a quick solution.

**Comment by sethrj** [comment]
@whokion I can restructure the input processing and push to your branch (or I can change it on develop and push to a new PR then you can rebase). 

WRT your interpolation: the `TwodGridCalculator` class (https://github.com/celeritas-project/celeritas/blob/develop/src/corecel/grid/TwodGridCalculator.hh) does bilinear interpolation on node-centered data. Is that what you need?

**Comment by whokion** [comment]
Thanks a lot!  Either way should be fine with me. Regarding the interpolation, it seems that TwodGridCalculator does what we need - I will digest it more carefully and use it as needed.

**Comment by sethrj** [comment]
To keep things simpler and spend less time resolving conflicts, maybe it would just be better to put `TODOs` in the class comments about the work to be done, and let the rest of the updates be resolved in further merge requests...

**Comment by sethrj** [comment]
OK, last thing for this merge maybe should be to rename FieldMap→RZFieldMap?

File: src/celeritas/field/RZMapField.hh [hunk 1]
```diff
@@ -53,18 +50,16 @@ class CMSMapField
  * Construct with the shared magnetic field map data (FieldMapData).
  */
 CELER_FUNCTION
-CMSMapField::CMSMapField(FieldMapRef const& shared) : shared_(shared) {}
+RZMapField::RZMapField(FieldMapRef const& shared) : shared_(shared) {}
 
 //---------------------------------------------------------------------------//
 /*!
- * Retrieve the magnetic field value for the given position.
+ * Retrieve the magnetic field value (in tesla) for the given position.
```
Comment (sethrj) [inline src/celeritas/field/RZMapField.hh]: The result should be "native" units, which I think is 0.001 T? The unit conversion on input may have been broken when I resolved the conflicts...

Comment (whokion) [reply src/celeritas/field/RZMapField.hh]: I thought that the return value is now in tesla as we dropped cerlitas::units::tesla here assuming that input field values are in tesla unit, which am confused - of course, we need a proper conversion for the field value between Geant4 (T=0.001) and celeritas (tesla=1e+4?) for the CMS integration.

Comment (sethrj) [reply src/celeritas/field/RZMapField.hh]: We should take the reader input in Tesla and convert in the Params constructor

**Review by sethrj — APPROVED** [review]

**Pull request merged**